src/query/boolean_query/block_wand.rs

```diff
@@ -195,6 +211,55 @@ pub fn block_wand(
     }
 }
 
+/// Specialized version of [`block_wand`] for a single scorer.
+/// In this case, the algorithm is simple and readable and faster (~ x3)
+/// than the generic algorithm.
+/// The algorithm behaves as follows:
+/// - While we don't hit the end of the docset:
+///   - While the block max score is under the `threshold`, go to the
+///     next block.
+///   - On a block, advance until the end and execute `callback``
+///     when the doc score is greater or equal to the `threshold`.
+pub fn block_wand_single_scorer(
+    mut scorer: TermScorer,
+    mut threshold: Score,
+    callback: &mut dyn FnMut(u32, Score) -> Score,
+) {
+    let mut doc = scorer.doc();
+    loop {
+        // We position the scorer on a block that can reach
+        // the threshold.
+        while scorer.block_max_score() < threshold {
+            let last_doc_in_block = scorer.last_doc_in_block();
+            if doc == TERMINATED {
+                return;
+            }
+            doc = last_doc_in_block + 1;
+            scorer.shallow_seek(doc);
+        }
+        // Seek will effectively load that block.
+        doc = scorer.seek(doc);
+        if doc == TERMINATED {
+            break;
+        }
+        loop {
+            let score = scorer.score();
+            if score > threshold {
+                threshold = callback(doc, score);
+            }
+            if doc >= scorer.last_doc_in_block() {
```

Comment:
This `>=` only triggers as == ?

Reply:
Indeed, we don't need the assert, this is pretty straightforward.